### ISSUE TITLE
Adding support for segmented sorting

### ISSUE THREAD

**Issue by ogreen** [cupy/cupy#6699]
### Description

cuPy's API currently does not support segmented sorting.
Segmented sorting is very useful for a wide range of data analytics (including sparse data applications and graph applications).
NVIDIA's CUB library has a highly tuned segmented sorting implementation starting as of CUB 1.15.  
Would it be possible to add support for this feature?

**Comment by leofang** [comment]
> NVIDIA's CUB library has a highly tuned segmented sorting implementation starting as of CUB 1.15.

This reminds me I wanna discuss the CUB support in CuPy. Currently CuPy uses the version seen at the compile time (which is decided by the CTK version). I will create a separate issue to discuss.

**Comment by leofang** [comment]
> I will create a separate issue to discuss.

See #6700.

**Comment by leofang** [comment]
@ogreen based on the offline discussion I believe the right signature for this need is:
```python
cupyx.segmented_sort(keys, offsets, values=None, ascending=True)
```
I have a working version to support it (see [here](https://github.com/leofang/cupy/tree/segmented_sort)), but it has a big drawback: Due to unconstrained key/value types it'd do a N^2 temperate specialization, which significantly slows down the compilation time of CuPy's CUB module (>4x slower) and inflates the binary size (~6x), for sm_86 **_alone_**. I don't think it's acceptable given that other CUB functions used in CuPy are much more lightweight.

What's the typical types when kv pairs need to be sorted? Would they usually have the same type? If so we can shrink N^2 to N. Alternatively, we can try writing our own segment-sort kernel based on CUB's `BlockRadixSort`, and see if NVRTC can jit it at runtime (we've done this for reduction). It'd not be a one-night project, though.

**Comment by emcastillo** [comment]
How about compiling it on the fly using `RawModule`?
(by @asi1024 )

**Comment by leofang** [comment]
@asi1024 @emcastillo I am making progress, slowly. If we could reach a conclusion for #6700, it seems possible to JIT-compile CUB's block sort kernel at runtime using Jitify. 

Currently, to make the snippet below work requires extra steps:
- Bump the bundled CUB version to 1.16+
- Bundle the bf16 headers
- Disable searching headers in `$CUDA_PATH/include`

Without the above steps the compilation would fail. Then, the next question would be to investigate what it takes to implement the segment support.

That said, it'd still be nice if @ogreen could confirm whether we really need such a wide support for k/v types.

```python
import cupy as cp

# need to use a more recent CUB (ex: 1.16.0) that includes the patch NVIDIA/cub#335,
# here I update CuPy's CUB bundle, and disabled searching in $CUDA_PATH/include.

# also, we need to copy/paste cuda_bf16.{h, hpp} to CuPy's header folder, as CUB
# needs it

code = """
// let Jitify handle these includes
#include <cstddef>  // for std::size_t
#include <type_traits>

namespace std {

template<class T>
struct is_pointer : std::false_type {};
 
template<class T>
struct is_pointer<T*> : std::true_type {};
 
template<class T>
struct is_pointer<T* const> : std::true_type {};
 
template<class T>
struct is_pointer<T* volatile> : std::true_type {};
 
template<class T>
struct is_pointer<T* const volatile> : std::true_type {};

}

#include <cupy/cub/cub/block/block_radix_sort.cuh>

__global__ void ExampleKernel(int* keys, int* offsets)
{
    // Specialize BlockRadixSort for a 1D block of 128 threads owning 4 integer items each
    typedef cub::BlockRadixSort<int, 128, 4> BlockRadixSort;

    // Allocate shared memory for BlockRadixSort
    __shared__ typename BlockRadixSort::TempStorage temp_storage;

    // Obtain a segment of consecutive items that are blocked across threads
    int thread_keys[4];

    // Collectively sort the keys
    BlockRadixSort(temp_storage).Sort(thread_keys);

    __syncthreads();
}
"""

mod = cp.RawModule(code=code,
                   options=('-std=c++14',),
                   jitify=True
                   )
mod.compile()
```

**Comment by leofang** [comment]
> * Bundle the bf16 headers

This is not allowed, so we need a workaround (NVIDIA/CUB#478).

**Comment by ogreen** [comment]
Yes. We do not to support k/v pairs. Consider the case where someone is building a sparse matrix via CSR such that edge has a weight. We need to be able to move the weights with the nnz (non-zero) entry.

**Comment by leofang** [comment]
> We do not to support k/v pairs.

I take it that you meant "we do need to"? And based on the use case k/v need to be of different types (ex: int for indices vs float for nnz entries)?

**Comment by leofang** [comment]
> > * Bundle the bf16 headers
> 
> This is not allowed, so we need a workaround ([NVIDIA/cub#478](https://github.com/NVIDIA/cub/issues/478)).

Better WAR: Just prepend this to trick CUB...
```c++
struct __nv_bfloat16 {};
```

**Comment by leofang** [comment]
@emcastillo @asi1024 Thoughts for discussion: We definitely don't want to merge [this version](https://github.com/cupy/cupy/issues/6699#issuecomment-1120348863) as it'd significantly increase the compilation time and binary size, making the redistribution on pypi.org/conda-forge almost impossible. However, I don't see a future-proof and easy way to get the necessary bits extracted from CUB.

What if we generate host code that calls CUB device algorithms (in this case, segmented sort) at runtime, and compile it with nvcc to a shared library (.so)?  The proposed solution would be akin to how we handle cuFFT callbacks, except that in this case I plan to not reply on Cython at all and instead I'd use ctypes to load the compiled shared library and get the function pointer directly. Upon completion the work would be two-fold:
1. A lightweight infrastructure for generating .so files that include the target CUDA kernels (fully instantiated)
2. For this case, support of segmented sort (equivalent to the linked WIP solution above).

Let me know if this is acceptable. Thanks.

**Comment by leofang** [comment]
Users should go use cuda.compute, which will become CuPy's default backend in the future (#9238).
